### ISSUE TITLE
[Help] GodotNative: accessing Rust nodes from Rust nodes - Liftime problem

### ISSUE THREAD

**Issue by mlange-42** [mlange-42/kohonen#14]
Trying to get a sibling node as Rust struct. Scene:

```
Node
 |- Kohonen
 '- Mapping
```

Trying to get the Rust Script attached to `Kohonen`, from `Mapping`:
```rust
use crate::Kohonen;
use gdnative::{Instance, Node2D, NodePath};

#[derive(gdnative::NativeClass)]
#[inherit(gdnative::Node2D)]
pub struct Mapping;

#[gdnative::methods]
impl Mapping {
    fn _init(_owner: gdnative::Node2D) -> Self {
        Mapping
    }

    #[export]
    fn _ready(&mut self, owner: gdnative::Node2D) {}

    #[export]
    pub fn _process(&mut self, owner: Node2D, _delta: f64) {}

    fn get_kohonen<'a>(&self, owner: &'a gdnative::Node2D) -> Option<&'a Kohonen> {
        let node = unsafe { owner.get_node(NodePath::from_str("../Kohonen")) };
        let koh: Option<&'a Kohonen> = node.and_then(|node| {
            Instance::<Kohonen>::try_from_base(node).map(|inst| inst.map(|koh, _| koh).unwrap())
        });
        koh
    }
}
```

Compiler errors:
```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
  --> kohvis\src\mapping_gd.rs:23:83
   |
23 |             Instance::<Kohonen>::try_from_base(node).map(|inst| inst.map(|koh, _| koh).unwrap())
   |                                                                                   ^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 23:74...
  --> kohvis\src\mapping_gd.rs:23:74
   |
23 |             Instance::<Kohonen>::try_from_base(node).map(|inst| inst.map(|koh, _| koh).unwrap())
   |                                                                          ^^^^^^^^^^^^
note: ...so that the expression is assignable
  --> kohvis\src\mapping_gd.rs:23:83
   |
23 |             Instance::<Kohonen>::try_from_base(node).map(|inst| inst.map(|koh, _| koh).unwrap())
   |                                                                                   ^^^
   = note: expected  `&kohonen_gd::Kohonen`
              found  `&kohonen_gd::Kohonen`
note: but, the lifetime must be valid for the lifetime `'a` as defined on the method body at 20:20...
  --> kohvis\src\mapping_gd.rs:20:20
   |
20 |     fn get_kohonen<'a>(&self, owner: &'a gdnative::Node2D) -> Option<&'a Kohonen> {
   |                    ^^
note: ...so that the expression is assignable
  --> kohvis\src\mapping_gd.rs:22:40
   |
22 |           let koh: Option<&'a Kohonen> = node.and_then(|node| {
   |  ________________________________________^
23 | |             Instance::<Kohonen>::try_from_base(node).map(|inst| inst.map(|koh, _| koh).unwrap())
24 | |         });
   | |__________^
   = note: expected  `std::option::Option<&'a kohonen_gd::Kohonen>`
              found  `std::option::Option<&kohonen_gd::Kohonen>`

error: aborting due to previous error
```

**Comment by mlange-42** [comment]
@adamreichold Any idea how to solve that? I was hoping to get the result with the lifetime of `owner`. Thanks!

**Comment by adamreichold** [comment]
> Any idea how to solve that? I was hoping to get the result with the lifetime of `owner`. Thanks!

I think the first attempt should be to drop the lifetime all together and pass `owner` by value. It already is a pointer to an object whose life time is managed by Godot. (Which is the reason why all actual accesses are unsafe.)

The second observation would be that you probably have to pass the full `Instance<Kohonen>` around as the Godot object part of it controls the life time even though you are only interested in the native class part.

(More specifically, the signature `fn Instance::map<F, U>(&self, op: F)` with ` F: for<'a>: FnOnce(&'a T, <T as NativeClass>::Base) -> U` informs you that you loose any control over the lifetime of the `&'a T` pass to you and hence its scope is effectively limited to your calling scope.)

Long story short: Rust lifetimes will probably not help much when interacting with objects managed by Godots runtime are they are currently exposed which explains the huge amount of unsafe functions in its API.

**Comment by adamreichold** [comment]
Meaning that your signature should probably just be ` fn get_kohonen(owner: gdnative::Node2D) -> Option<Instance<Kohonen>>`. (I don't think you actually need `&self` here?)

**Comment by mlange-42** [comment]
@adamreichold Thanks!
So far, I solved it by a function taking a closure, so I don't need to care for lifetimes:

```
trait KohonenUser2D {
    fn with_kohonen<F>(mut owner: gdnative::Control, path: &str, mut fun: F)
    where
        F: FnMut(&mut gdnative::Control, &Kohonen),
    {
        let node = unsafe { owner.get_node(NodePath::from_str(path)) };
        node.and_then(|node| {
            Instance::<Kohonen>::try_from_base(node)
                .map(|inst| inst.map(|koh, _| fun(&mut owner, koh)))
        });
    }
}
```

**Comment by adamreichold** [comment]
> @adamreichold Thanks!
> So far, I solved it by a function taking a closure, so I don't need to care for lifetimes:

Closure-based are often the last resort when fixing lifetime issue due to the implicit HKTB on the closure types `FnOnce`, `Fn` and `FnMut`. Keep in mind that fetch the node via `get_node` is not particularly cheap and even `NodePath::from_str` will allocate.
